WIP: Node split

Lyla-Fischer opened the pull request:
Addresses #12 

I'm a little bit worried about "just" passing tests, without a more comprehensive test suite, considering that I was very much just fixing bugs as the tests revealed them, but this passes tests.

I put this on top of the string escaping, because I didn't want to deal with the inevitable conflicts.

Lyla-Fischer committed:
Add a couple of comments to _types

Lyla-Fischer committed:
Escape quotes in strings

Lyla-Fischer committed:
Move Node classmethods to node_type_registry.

ahopkins commented on pydiggy/node.py:
```diff
@@ -136,24 +123,42 @@ def __new__(cls, name, bases, attrs, **kwargs):
 
 
 class Node(metaclass=NodeMeta):
-    uid: int
```
Why remove this? It will be missing from the `annotations` and not added as a property, unless it is somewhere else now?

Lyla-Fischer replied:
I just removed the class variable. The instance variable self.uid is still declared in `__init__`, and therefore all of the instances still have it. I removed it because I thought it was confusing, and it was never actually used.

ahopkins replied:
I think it should stay so that the `uid` is declared as a type annotation.

Lyla-Fischer replied:
Okay. I put it back, and added a comment so that future me will not be confused (if that's okay).

ahopkins commented on pydiggy/node.py:
```diff
@@ -136,24 +123,42 @@ def __new__(cls, name, bases, attrs, **kwargs):
     def __init_subclass__(cls, is_abstract: bool = False) -> None:
         if not is_abstract:
-            cls._register_node(cls)
+            from pydiggy.node_type_registry import NodeTypeRegistry
```
Why the import inside the method?

Lyla-Fischer replied:
Avoiding circular imports. Short of doing more extensive refactoring, putting imports inside of implementations is the way to do it. 

I think that the only way to avoid this would be to have NodeTypeRegistry be in the same file as Node, or to remove the auto-registration.

ahopkins replied:
Got it. I would rather them be in the same file than nesting the imports. And, I see the auto-registration as a necessary feature. Maybe it can be turned off in the future, but it makes the end API super simple and intuitive to use (which was the goal).

ahopkins commented on pydiggy/node.py:
```diff
@@ -642,25 +356,6 @@ def _explode(
         return obj
 
-    @classmethod
-    def create(cls, **kwargs) -> Node:
```
Doesn't this break the API?

```python
class MyNode(Node):
    pass

my_node = MyNode.create(foo="bar")
```

Lyla-Fischer replied:
Ah... yes, it does. Before I run off an fix it, though: Is there a reason that this logic isn't in the `__init__` method? That's where I'm used to putting constructor-type logic.

ahopkins replied:
So that in the future it might also be able to handle an `awaitable` interface:

```python
my_node = await MyNode.create(foo="bar")
```

No way to achieve this with `__init__`.

Lyla-Fischer committed:
Put back the async constructor.

Lyla-Fischer committed:
Move NodeTypeRegistry to node.py.

Lyla-Fischer committed:
Format code with black.

Lyla-Fischer commented:
Addressed some of the comments. I would still like to add more tests before merging this in, but one of the major contributors to untested code metrics is the existence of `Node.save()` and connections.py, which do more than just serialization/de-serialization, which (I think, based off of other discussions) is the focus of the current library. 

I don't know if you want to improve the code coverage as part of this pull request or not, but I'm willing to wait for a merge to get some of that done, and make it more safe.

Lyla-Fischer committed:
Add Node.uid back, with comment.

ahopkins commented on pydiggy/node.py:
```diff
@@ -136,6 +136,10 @@ class Node(metaclass=NodeMeta):
 
+    # uid is not used as a class variable, it is part of the required
+    # schema of a Node, and is thus part of the superclass.
+    uid : int
```
Nice documentation :muscle:
